ujt/utils.py

```diff
@@ -131,3 +86,44 @@ def get_highest_collapsed_virtual_node_name(node_name):
         highest_collapsed_name = current_name
 
     return highest_collapsed_name
+
+
+def proto_list_to_name_map(proto_list):
+    return {proto.name: proto for proto in proto_list}
+
+
+def get_existing_uuid(elements):
+    return elements[0]["data"]["id"].rsplit("#", 1)[1]
+
+
+def get_all_node_names_within_virtual_node(
+        virtual_node_name,
+        node_name_message_map,
+        virtual_node_map):
+    node_names = set()
+    node_frontier = deque(virtual_node_map[virtual_node_name].child_names)
+
+    while node_frontier:
+        current_node_name = node_frontier.popleft()
+        if current_node_name in node_name_message_map:
+            node_names.add(current_node_name)
+        if current_node_name in virtual_node_map:
+            for child_name in virtual_node_map[current_node_name].child_names:
+                node_frontier.append(child_name)
+
+    return node_names
+
+
+def string_to_dict(stringified_dict):
+    return json.loads(stringified_dict)
+
+
+def get_latest_tapped_element(tap_node, tap_edge):
+    # is this a hack?
```

Comment:
it seems fine... but maybe the method should be renamed? Since it is only comparing between two things

Reply:
tap_node should always be updated to the latest tapped node (among all nodes), and same with tap_edge, but I agree it could be clearer. Something like `get_recent_tapped_element(tap_node, tap_edge)`?